### ISSUE TITLE
Swin-Transformer Loss对齐实验

### ISSUE THREAD

**Issue by rentainhe** [Oneflow-Inc/swin-transformer#2]
# Swin-Transformer Loss对齐实验结果汇总
## 实验
- 使用torch的dataloader，保证相同输入的情况下，用SGD和AdamW优化器分别迭代1000个iters的结果: [__experiment results__](https://github.com/Oneflow-Inc/swin-transformer/issues/2#issuecomment-999264174)
- 使用torch的dataloader，保证相同输入的情况下，用SGD优化器，添加`clip-grad`: [__experiment results__](https://github.com/Oneflow-Inc/swin-transformer/issues/2#issuecomment-997701836)
- 开启`cudnn_determinstic=True`，`model.eval()`的条件下，无论是`torch与oneflow比较`还是`使用oneflow自身跑两次实验`都可以观察到以下结论: [__experiment results__](https://github.com/Oneflow-Inc/swin-transformer/issues/2#issuecomment-999186666)
  - 使用随机数据，AdamW优化器的Loss可以完全对齐
  - 使用真实数据，AdamW优化器的Loss不能对齐
  - 使用随机数据，将AdamW优化器的学习率调大之后，Loss不能对齐: [__experiment results__](https://github.com/Oneflow-Inc/swin-transformer/issues/2#issuecomment-999400008)
- 对flowvision中的模型进行抽样，发现`vit`, `mlp-mixer`一组的模型在AdamW下的Loss都不能完全对齐:  [__experiment results__](https://github.com/Oneflow-Inc/swin-transformer/issues/2#issuecomment-999618005)
- 用拼的 layernorm 替换内置的 layernorm 并且将 mean 和 std 都替换成 sum后，在AdamW下Swin的Loss可以完全对齐: [__experiment results__](https://github.com/Oneflow-Inc/swin-transformer/issues/2#issuecomment-1000001409)
- 其他情况不变，将`gelu`算子替换成`relu`后, 设置`layernorm`的`elementwise_affine= False`, 关闭`FAST_MATH`, 在AdamW优化器下的Loss对齐情况: [__experiment results__](https://github.com/Oneflow-Inc/swin-transformer/issues/2#issuecomment-1000021308)
  - `batch_size=8`的时候500个iter下loss基本可以对齐
  - `batch_size=16`的时候在大约200个iter后loss不对齐
- 关闭`FAST_MATH`，去掉`LayerNorm`，使用修复的GELU后的Loss对齐实验: [__experiment results__](https://github.com/Oneflow-Inc/swin-transformer/issues/2#issuecomment-1000631130)
  - AdamW优化器, lr = 1e-6：可以对齐
  - AdamW优化器, lr = 1e-5：可以对齐
  - AdamW优化器, lr = 1e-4：可以对齐
  - AdamW优化器, lr = 1e-3：OneFlow出现NAN
- 打开`mixup`，关闭`clip-grad`，使用`SoftTargetCE`，OneFlow收敛速度明显慢于torch: [__experiment results__](https://github.com/Oneflow-Inc/swin-transformer/issues/2#issuecomment-1001159164)
  - 检查发现`Mixup`中有部分Inplace操作没有替换，导致输入的label改变了，但是input没有改变，正在fix
- `2021.12.28`: 基于https://github.com/Oneflow-Inc/oneflow/pull/7103, 重新测试了`GELU`以及`LayerNorm`，发现LayerNorm的Loss无法对齐，GELU的可以对齐: [__experiment results__](https://github.com/Oneflow-Inc/swin-transformer/issues/2#issuecomment-1001831969)
- `2021.12.29`: 基于 https://github.com/Oneflow-Inc/oneflow/pull/7103, 使用拼的`LayerNorm`, 加载与torch相同初始化的模型后，并在训练的时候开启clip-grad和mixup，在CIFAR100数据集上精度仅差0.3%: [__experiment results__](https://github.com/Oneflow-Inc/swin-transformer/issues/2#issuecomment-1002546094)
- `2021.12.29`: 正在基于 https://github.com/Oneflow-Inc/oneflow/tree/dev_debug_swin_transformer_gr 分支进行实验
- `2021.12.30`: 基于https://github.com/Oneflow-Inc/oneflow/tree/dev_debug_swin_transformer_gr 分支和`master`分支，载入相同的初始化权重进行训练

## 遇见的问题以及相关解决的PR
| Problems | Fixed or not | Fixed PR |
|:---:|:---:|:---:|
| `LayerNorm`可能存在问题 | Not Fixed | - |
| 内置的`GELU`可能存在问题 | Fixed | https://github.com/Oneflow-Inc/oneflow/pull/7103 |
| `Mixup`中的inplace操作替换 | Fixed | https://github.com/Oneflow-Inc/vision/pull/80 |

## Training Experiments
- 在默认超参的情况下的实验

| Model | Optimizer | Dataset | Framework | Acc@1 | Log |
|:---:|:---:|:---:|:---:|:---:|:---:|
| Swin-Tiny | AdamW | CIFAR100 | Pytorch | 76.77% | - |
| Swin-Tiny | AdamW | CIFAR100 | OneFlow | 67.29% | - |
| Swin-Tiny 将 LayerNorm替换为python层拼接的Module | AdamW | CIFAR100 | OneFlow | 66.44% | [log](https://github.com/Oneflow-Inc/swin-transformer/files/7770298/log_rank0.txt)
| Swin-Tiny 将激活函数替换为ReLU | AdamW | CIFAR100 | Pytorch | - | - |
| Swin-Tiny 将激活函数替换为ReLU | AdamW | CIFAR100 | OneFlow | 66.39% | [log](https://github.com/Oneflow-Inc/swin-transformer/files/7772051/log_rank0.1.txt) |
| Swin-Tiny 使用拼的LayerNorm + ReLU | AdamW | CIFAR100 | OneFlow | - | - |

#### 基于 https://github.com/Oneflow-Inc/oneflow/pull/7103 编译的环境进行的实验，关闭`FAST_MATH`，修复`GELU`计算
- 关闭`Mixup`，关闭`clip-grad`，只用`CrossEntropyLoss`的情况下，使用内置的`LayerNorm`

| Model | Optimizer | Dataset | Framework | Acc@1 | Log |
|:---:|:---:|:---:|:---:|:---:|:---:|
| Swin-Tiny + ReLU | AdamW | CIFAR100 | Pytorch | 75.95% | [log](https://github.com/Oneflow-Inc/swin-transformer/files/7774892/log_rank0.txt) |
| Swin-Tiny + ReLU | AdamW | CIFAR100 | OneFlow | 74.74% |  [log](https://github.com/Oneflow-Inc/swin-transformer/files/7774906/log_rank0.2.txt) |
| Swin-Tiny + GELU | AdamW | CIFAR100 | Pytorch   | 75.41% |  [log](https://github.com/Oneflow-Inc/swin-transformer/files/7776837/log_rank0.txt) |
| Swin-Tiny + GELU | AdamW | CIFAR100 | OneFlow | 74.45% | [log](https://github.com/Oneflow-Inc/swin-transformer/files/7775357/log_rank0.3.txt)|

- 基于https://github.com/Oneflow-Inc/vision/pull/80, 打开`Mixup`，关闭`clip-grad`，使用`SoftTargetCE`, `label-smooth = 0`

| Model | Optimizer | Dataset | Framework | Acc@1 | Log |
|:---:|:---:|:---:|:---:|:---:|:---:|
| Swin-Tiny with error mixup | AdamW | CIFAR100 | OneFlow | 67.80% | [log](https://github.com/Oneflow-Inc/swin-transformer/files/7778079/error_mixup.txt)|
| Swin-Tiny with fixed mixup | AdamW | CIFAR100 | OneFlow | 75.68% | [log](https://github.com/Oneflow-Inc/swin-transformer/files/7779575/fix_mixup.txt) |
| Swin-Tiny with mixup | AdamW | CIFAR100 | Pytorch | 77.54% | [log](https://github.com/Oneflow-Inc/swin-transformer/files/7780532/torch_with_mixup.txt)|
